win32ss/user/user32/misc/logon.c

```diff
@@ -83,6 +84,69 @@ Logon(BOOL IsLogon)
                          sizeof(*LogonRequest));
 }
 
+/* Win: LoadPreloadKeyboardLayouts */
+BOOL IntLoadPreloadKeyboardLayouts(VOID)
+{
+    UINT nNumber, uFlags;
+    DWORD cbValue, dwType;
+    WCHAR szNumber[32], szValue[9];
+    HKEY hPreloadKey;
+    BOOL ret = FALSE;
+    HKL hKL, hDefaultKL = NULL;
+
+    if (RegOpenKeyW(HKEY_CURRENT_USER,
+                    L"Keyboard Layout\\Preload",
+                    &hPreloadKey) != ERROR_SUCCESS)
+    {
+        return FALSE;
+    }
+
+    for (nNumber = 1; nNumber <= 1000; ++nNumber)
+    {
+        StringCchPrintfW(szNumber, _countof(szNumber), L"%u", nNumber);
+
+        cbValue = sizeof(szValue);
+        if (RegQueryValueExW(hPreloadKey,
+                             szNumber,
+                             NULL,
+                             &dwType,
+                             (LPBYTE)szValue,
+                             &cbValue) != ERROR_SUCCESS)
+        {
+            break;
+        }
+
+        if (dwType != REG_SZ)
+            continue;
+
+        if (nNumber == 1)
+            uFlags = KLF_SUBSTITUTE_OK | KLF_ACTIVATE | KLF_RESET;
+        else
+            uFlags = KLF_SUBSTITUTE_OK | KLF_NOTELLSHELL | KLF_REPLACELANG;
+
+        hKL = LoadKeyboardLayoutW(szValue, uFlags);
+        if (hKL)
+        {
+            ret = TRUE;
```

Comment:
It looks like one could get rid of that "ret" variable, and: 1. initialize hKL originally to NULL (where it's defined), and 2. replace all those "ret" checks by suitable checks on the value of hKL; finally, 3. replacing `return ret;` by `return (hKL != NULL);`

Reply:
> I can't agree that. How to check `hKL` without assignment?

> 1. initialize hKL originally to NULL (where it's defined)

i.e. `HKL hKL = NULL, hDefaultKL = NULL;` that is at the top of the function. Then the hKL variable is always correctly defined: to NULL if nothing has been loaded, or to a successfully loaded layout.

Reply:
Difficult to see your opinion.

- There can be zero, single or multiple layout(s).
- `LoadKeyboardLayoutW` can be called in multiple times. Not just once.

Reply:
> LoadKeyboardLayoutW can be called in multiple times. Not just once.

Oh indeed! So you would return TRUE if at least one LoadKeyboardLayoutW call succeed? (even if some others didn't?) If so, then my proposition doesn't make sense, and indeed you should keep the separate `ret` variable.

Reply:
The return value is not important because the caller ignores it.

Reply:
Then why is there a `ret` variable altogether? The IntLoadPreloadKeyboardLayouts should then just be "void" :smile:

Reply:
Now return void. OK.